sage/example-transition/src/rules/hero_jam.rs

```diff
@@ -1,91 +1,119 @@
-use crate::{asset, asset::hero_jam, rules::RuleVerifier};
-
-use ajuna_primitives::{
-	asset_manager::{AssetInspector, AssetManager},
-	chain_inspector::ChainInspector,
+use crate::asset::{
+	hero_jam::{AssetType, HeroJamAsset, StateType},
+	Asset, AssetId,
 };
-use sage_api::Error;
 
-use crate::asset::AssetVariant;
-use frame_support::pallet_prelude::{Decode, Encode, MaxEncodedLen, TypeInfo};
-use parity_scale_codec::Codec;
-use sp_runtime::traits::{BlockNumber as BlockNumberT, Member};
-use std::marker::PhantomData;
+use ajuna_primitives::{asset_manager::AssetInspector, chain_inspector::ChainInspector};
+use sage_api::RuleError;
+use sp_runtime::traits::BlockNumber as BlockNumberT;
+
+pub const ASSET_NOT_FOUND: u8 = 200;
+pub const ASSET_NOT_HERO_JAM: u8 = 201;
+
+pub const ASSETS_NOT_ALL_SAME_TYPE: u8 = 100;
+pub const ASSETS_NOT_ALL_SAME_STATE: u8 = 101;
+pub const ASSET_CANNOT_STATE_CHANGE: u8 = 102;
+pub const ASSET_HERO_NOT_IN_ACCOUNT: u8 = 103;
+pub const ASSET_HERO_ALREADY_IN_ACCOUNT: u8 = 104;
+
+fn try_cast_to_hero_jam_asset<AccountId, BlockNumber, Inspector>(
+	asset_id: &AssetId,
+) -> Result<HeroJamAsset<BlockNumber>, RuleError>
+where
+	Inspector: AssetInspector<AccountId = AccountId, AssetId = AssetId, Asset = Asset<BlockNumber>>,
+{
+	if let Ok(asset) = Inspector::get_asset(asset_id) {
+		if let Ok(hero_jam_asset) = HeroJamAsset::try_from(asset) {
+			Ok(hero_jam_asset)
+		} else {
+			Err(RuleError::Other { error: ASSET_NOT_HERO_JAM })
+		}
+	} else {
+		Err(RuleError::Other { error: ASSET_NOT_FOUND })
+	}
```

Comment:
I prefer to avoid double nesting, and do something like this, and I think we can in general make this more elegant:

```rust
let asset = Inspector::get_asset(asset_id).ok_or(RuleError::Other { error: ASSET_NOT_FOUND })?;
asset.try_into().map_err(|_| RuleError::Other { error: ASSET_NOT_HERO_JAM })
```